fix #61: add macro nutrient tracking

conrad69 opened the pull request:
This enhances the app so users can add macro nutrients to the products they add. The macro nutrients for each entry are listed in the overview and below the overview heading the total macros are shown also. Products added from the OpenFoodFacts search also contain macros - if OFF provides them. The statisitics show a graph view for every macro below the energy graph view - carbs & sugar and fat & saturated fat each share a graph view.

I was not able to test how this enhacement works when updating an existing version of the app, I guess there need changes to be done to some database stuff, so that the upgrade works.

Cliveptr commented:
Do you have any plans to finish (test) this? The lack of macro tracking is the only thing preventing me from being able to use the app :'(

conrad69 commented:
Hi, finally I had some time to look into this again. I merged the recent changes and refactored it a little. I also fixed the database upgrade issue, now it should be possible to upgrade an existing installation without loosing the data - I only tested this on my phone though.
If the changes make it into the official app, it would probably be a good idea to give users the ability to edit the product database so they can add the other nutrition info to existing entries. I think I will try to implement something like that in the future.

cloehle commented:
Just want to add that Ive been using this branch for a while and think it works very well and is a good addition to the app. Exactly what I needed.
Modifying existing entries to add macros would be nice, but modifying existing entries in general would be nice to have
Thank you very much for your work @conrad69

Yky commented:
I have also been using this branch for a many months and agree with @cloehle it works very well. Thanks @conrad69 

Would be great to get a review done for this branch and get it merged.

conrad69 commented:
Hi,
I reworked the macro nutrient tracking and enhanced it to now be able to track 36 nutrients:  carbs, sugar, protein, fat, satFat, salt, fiber, vitaminA_retinol, betaCarotin, vitaminD, vitaminE, vitaminK, thiamin_B1, riboflavin_B2, niacin, vitaminB6, folat, pantothenacid, biotin, cobalamin_B12, vitaminC, natrium, chlorid, kalium, calcium, phosphor, magnesium, eisen, jod, fluorid, zink, selen, kupfer, mangan, chrom, molybdaen

In the newly added Settings-Activity users can choose which nutrient to track. For every nutrient and the calories one can set a daily goal and whether one should surpass this goal or not. In the overview activity the goals are shown as progress bars. The following screenshot shows an example, where 4 nutrients have goals set. The sugar goal shall not be surpassed and it isn't, so the progress bar is green. The goal of 150g protein shall be surpassed every day, so the progress bar is orange, since that isn't the case. Salt and fiber progress bars show the behaviour when a goal is surpassed.

![FoodTracker_Overview](https://user-images.githubusercontent.com/9432071/135494266-67836184-ebfe-4750-9ce9-6b0e1127ad82.jpg)

This fixes #61 , #85 and #73 in a way.

conrad57 commented:
Without me noticing one of my previous commits introduced a bug where the settings activity would not allow to enable any nutrient tracking. This is fixed now. 
I furthermore removed some unused methods which were introduced in an earlier version of the tracking feature.

cloehle commented:
Very cool, I'll try it out and report back.

Kamuno reviewed (changes requested):
So I assume db version should be 3. Also I just tested this build and it gives me a NPE every time I choose something from the search results.

It looks good so far, though.

conrad57 commented on app/src/main/java/org/secuso/privacyfriendlyfoodtracker/database/ApplicationDatabase.java:
```diff
@@ -72,14 +73,46 @@ public void onCreate(@NonNull SupportSQLiteDatabase db) {
+    /***
+     * Migrate database so it also contains micro nutriments.
+     */
+    static final Migration MIGRATION_2_3 = new Migration(2, 3) {
```
I think you are right about the database version. I did not look into this for a long time though.
I can look into the npe in like one or two weeks. Do you have more information on the exception?

ReIaxo replied:
Hi, I also took a look and fixed some bugs: 
- At some places product.key was used instead of offKey which leads to Zero values in database products
- The click listener in search products was weird implemented. If there was a reason for it I don't get it. I cleaned it up a little which fixed the crash mentioned above.
How can i contribute my changes? This will be my first contribution to an Open Source Project on GitHub.
I have done the changes in the forked repositpory of @conrad57 an have them only local on my machine.
@Kamuno  And there are merge conflicts. Which branch do I have to merge into this to resolve the conflicts?
